Mixpanel/MixpanelExceptionHandler.m

```diff
@@ -93,88 +108,66 @@ - (void)addMixpanelInstance:(Mixpanel *)instance {
     [self.mixpanelInstances addObject:instance];
 }
 
-void MPSignalHandler(int signal)
-{
+void MPSignalHandler(int signal, struct __siginfo *info, void *context) {
+    MixpanelExceptionHandler *handler = [MixpanelExceptionHandler sharedHandler];
+
     int32_t exceptionCount = OSAtomicIncrement32(&UncaughtExceptionCount);
-    if (exceptionCount > UncaughtExceptionMaximum)
-    {
-        return;
+    if (exceptionCount <= UncaughtExceptionMaximum) {
+        NSArray *callStack = [MixpanelExceptionHandler backtrace];
+        NSDictionary *userInfo = @{UncaughtExceptionHandlerSignalKey: @(signal),
+                                   UncaughtExceptionHandlerAddressesKey: callStack};
+        NSException *exception = [NSException exceptionWithName:UncaughtExceptionHandlerSignalExceptionName
+                                                         reason:[NSString stringWithFormat:@"Signal %d was raised.", signal]
+                                                       userInfo:userInfo];
+
+        [handler mp_handleUncaughtException:exception];
     }
 
-    NSMutableDictionary *userInfo =
-    [NSMutableDictionary
-     dictionaryWithObject:[NSNumber numberWithInt:signal]
-     forKey:UncaughtExceptionHandlerSignalKey];
-
-    NSArray *callStack = [MixpanelExceptionHandler backtrace];
-    [userInfo
-     setObject:callStack
-     forKey:UncaughtExceptionHandlerAddressesKey];
-
-    [MixpanelExceptionHandler performSelectorOnMainThread:@selector(mp_handleUncaughtException:)
-                              withObject:[NSException
-                                          exceptionWithName:UncaughtExceptionHandlerSignalExceptionName
-                                          reason:
-                                          [NSString stringWithFormat:
-                                           NSLocalizedString(@"Signal %d was raised.", nil),
-                                           signal]
-                                          userInfo:
-                                          [NSDictionary
-                                           dictionaryWithObject:[NSNumber numberWithInt:signal]
-                                           forKey:UncaughtExceptionHandlerSignalKey]] waitUntilDone:YES];
+    struct sigaction prev_action = handler.prev_signal_handlers[signal];
+    if (prev_action.sa_flags & SA_SIGINFO) {
```

Comment:
can you explain this part?

Reply:
yeah if `SA_SIGINFO` is set in the flags it uses `sa_sigaction` as the registered function, otherwise it's `sa_handler`. it's actually in the same location in memory because it's a `union` but the signature is different